L73-java-binderchannel/wireformat.md

```diff
@@ -25,38 +25,60 @@ There are two types of transactions, control and stream.
 For general transport lifecycle, using one of 1000 reserved transaction codes.
 
 *   Setup Transport
-    *   Transaction code: BinderTransport.SETUP_TRANSPORT
+    *   Transaction code: BinderTransport.SETUP_TRANSPORT (1)
     *   From ClientTransport to host binder, and from ServerTransport to client
         binder.
 *   Shutdown Transport
-    *   Transaction code: BinderTransport.SHUTDOWN_TRANSPORT
+    *   Transaction code: BinderTransport.SHUTDOWN_TRANSPORT (2)
     *   From ClientTransport to server binder, or ServerTransport to client
         binder.
 *   Acknowledge bytes.
-    *   Transaction code: BinderTransport.ACKNOWLEDGE_BYTES
+    *   Transaction code: BinderTransport.ACKNOWLEDGE_BYTES (3)
     *   From either transport to its peer's binder. Reports reception of
         transaction data for flow control.
 *   Ping
-    *   Transaction code: BinderTransport.PING
-    *   From ClientTransport to server binder.
+    *   Transaction code: BinderTransport.PING (4)
+    *   Currently only sent from ClientTransport to server binder.
 *   Ping Response:
-    *   Transaction code: BinderTransport.PING_RESPONSE
-    *   From SeverTransport to client binder, in response to a ping.
+    *   Transaction code: BinderTransport.PING_RESPONSE (5)
+    *   From SeverTransport to client binder or ClientTransport to server binder,
+        in response to a ping message.
+
+If either side of a transport receives an unrecognized control transacton,
+it will shutdown the transport gracefully.
 
 #### Transaction Data
 
 ```
-version = int; (* Currently always 1 *)
+major version = int;
+minor version = int;
+protocol extension flags = int;
+version extras = minor version, protocol extension flags
 num bytes = int;
 ping id = int;
+version info = major version, [version extras] (* If major version is at least 2. *)
+
 
-setup transport transaction = version, binder;
+setup transport transaction = version info, binder;
```

Comment:
As discussed separately, we could add `[version extras]` after `binder` and avoid the need to bump the protocol version.

Reply:
I went with just putting the protocol extension flags after binder, since we discussed dropping the minor version.